SO-5468 Filter resources by settings

AAAlinaaa opened the pull request.

AAAlinaaa committed:
feat(resources): Support filtering by settings parameters

AAAlinaaa committed:
Merge branch '8.x' into issue/SO-5468-filter_resources_by_settings

Conflicts:
	core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/BaseResourceSearchRequest.java
	core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/BaseResourceSearchRequestBuilder.java

AAAlinaaa committed:
fix(resource-settings) filter by options

AAAlinaaa committed:
feat(resources): allow filtering resources by properties

AAAlinaaa committed:
(resources): fix enum description

AAAlinaaa committed:
chore(resource): ogranize import statements

cmark commented on core/com.b2international.snowowl.core.rest/src/com/b2international/snowowl/core/rest/domain/BaseResourceRestSearch.java:
```diff
@@ -42,6 +42,9 @@ public abstract class BaseResourceRestSearch extends ObjectRestSearch {
 	private List<String> owner;
 	
+	@Parameter(description = "One or more property names or property-value pairs separated by a dot.")
+	private List<String> properties;
```
Rename to settings, as the field we are filtering by is the `settings` field.

cmark commented on core/com.b2international.snowowl.core.rest/src/com/b2international/snowowl/core/rest/resource/ResourceRestService.java:
```diff
@@ -65,6 +65,7 @@ public Promise<Resources> searchByGet(@ParameterObject final ResourceRestSearch
 			.filterByStatus(params.getStatus())
 			.filterByUrls(params.getUrl())
 			.filterByOwner(params.getOwner())
+			.filterByProperties(params.getProperties())
```
Same here, rename all filterByProperties method to `filterBySettings`.

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/BaseResourceSearchRequest.java:
```diff
@@ -81,6 +81,11 @@ public enum OptionKey {
+		/**
+		 * Search resources by a concrete property and value pair in settings
+		 */
+		PROPERTIES
```
And this as well.

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/BaseResourceSearchRequest.java:
```diff
@@ -98,6 +103,21 @@ protected final Expression prepareQuery(RepositoryContext context) {
 		}
 
+		if (containsKey(OptionKey.PROPERTIES)) {
+			final Collection<String> properties = getCollection(OptionKey.PROPERTIES, String.class);
```
A few example test cases would be nice to verify the functionality. Also, not sure if dot is the best separator for key and value pairs. What happens with nested settings, like `settings.nested.prop.field`.

AAAlinaaa committed:
test(resources): Add filter by settings test cases

AAAlinaaa committed:
fix(resources): Reserve dot to define nested property names, use # to...

... separate property names from values, defer to using the "settings"
filter key

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/BaseResourceSearchRequest.java:
```diff
@@ -98,6 +103,21 @@ protected final Expression prepareQuery(RepositoryContext context) {
+			properties.forEach( property -> {
+				if (property.contains("#")) {
+					final String propertyName = property.split("\\#")[0];
+					final String propertyValue = property.substring(propertyName.length() + 1, property.length());
```
If I specify `key#` without value this will throw an Exception, please handle that and create a test case for it.

AAAlinaaa committed:
SO-5468: Throw error on invalid settings filter argument formatting

cmark reviewed (approved):
Looks good!
